test/refunder.ts

```diff
@@ -331,6 +344,81 @@ describe("Refunder", function() {
 			const balanceAfter = await ethers.provider.getBalance(userAddress);
 			expect(balanceAfter.lt(balanceBefore), 'User was refunded').to.be.ok;
 		});
+
+		it('NOT approved user should NOT be refunded', async () => {
+
+			const funcIdAsBytes = generateFuncIdAsBytes('setGreeting(string)');
+			const validationFuncIdAsBytes = generateFuncIdAsBytes('isApproved(address)');
+
+			let res = await refunder.updateRefundable(greeter.address, funcIdAsBytes, true, permitter.address, validationFuncIdAsBytes);
+			await res.wait();
+
+			const text = 'Hello, Tester!';
+			const hexString = strToHex(text);
+			const args = ethers.utils.arrayify(hexString);
+
+			await expect(refunder.connect(addr1).relayAndRefund(greeter.address, funcIdAsBytes, args)).to.be.revertedWith(NOT_PERMITTED);
+		});
+
+		it('After permit user should be refunded', async () => {
+
+			const funcIdAsBytes = generateFuncIdAsBytes('setGreeting(string)');
+			const validationFuncIdAsBytes = generateFuncIdAsBytes('isApproved(address)');
+
+			await permitter.updateRefundableUser(addr1.address, true);
+			
+			let res = await refunder.updateRefundable(greeter.address, funcIdAsBytes, true, permitter.address, validationFuncIdAsBytes);
+			await res.wait();
+
+			const balanceBefore = await ethers.provider.getBalance(addr1.address);
+			
+			const text = 'Hello, Tester!';
+			const hexString = strToHex(text);
+			const args = ethers.utils.arrayify(hexString);
+			
+			res = await refunder.connect(addr1).relayAndRefund(greeter.address, funcIdAsBytes, args);
+
+			let txReceipt = await res.wait();
+			let txCost = res.gasPrice.mul(txReceipt.gasUsed)
+			
+			const balanceAfter = await ethers.provider.getBalance(addr1.address);
+			expect(balanceAfter.lt(balanceBefore), 'Sender was over refunded').to.be.ok;
+			
+			const cost = res.gasPrice.mul(txReceipt.cumulativeGasUsed);
+			const reimbursement = balanceAfter.sub(balanceBefore.sub(cost));
+			expect(reimbursement.lt(cost), 'Sender was over refunded').to.be.ok;
+			
+			expect(txReceipt.events, 'No events are emitted').to.be.ok;
+			expect(txReceipt.events[0].event, 'Invalid event name').to.be.eq('RelayAndRefund');
+
+			const difference = balanceBefore.sub(balanceAfter);
+			const percent = 5;
+			let percentFrom = getXPercentFrom(txCost, percent);
+			expect(difference.lt(percentFrom), `User was refunded less than ${100 - percent}%`);
+
+			res = await greeter.greet();
+			expect(res).to.be.equal(text);
```

Comment:
This logic seems that it is duplicated a couple of times. Lets extract it to a separate function